network/default.go

```diff
@@ -0,0 +1,336 @@
+				Type:      router.AdvertType(pbAdvert.Type),
+				Timestamp: time.Unix(0, pbAdvert.Timestamp),
+				TTL:       time.Duration(pbAdvert.Ttl),
+				Events:    events,
+			}
+
+			if err := n.Router.Process(advert); err != nil {
+				log.Debugf("Network failed to process advert %s: %v", advert.Id, err)
+				continue
+			}
+		}
+	}
+}
+
+// advertise advertises routes to the network
+func (n *network) advertise(client transport.Client, advertChan <-chan *router.Advert) {
+	for {
+		select {
+		// process local adverts and randomly fire them at other nodes
+		case advert := <-advertChan:
+			// create a proto advert
+			var events []*pb.Event
+			for _, event := range advert.Events {
+				route := &pb.Route{
+					Service: event.Route.Service,
+					Address: event.Route.Address,
+					Gateway: event.Route.Gateway,
+					Network: event.Route.Network,
+					Link:    event.Route.Link,
+					Metric:  int64(event.Route.Metric),
+				}
+				e := &pb.Event{
+					Type:      pb.EventType(event.Type),
+					Timestamp: event.Timestamp.UnixNano(),
+					Route:     route,
+				}
+				events = append(events, e)
+			}
+			pbAdvert := &pb.Advert{
+				Id:        advert.Id,
+				Type:      pb.AdvertType(advert.Type),
+				Timestamp: advert.Timestamp.UnixNano(),
+				Events:    events,
+			}
+			body, err := proto.Marshal(pbAdvert)
+			if err != nil {
+				// TODO: should we bail here?
+				log.Debugf("Network failed to marshal message: %v", err)
+				continue
+			}
+			// create transport message and chuck it down the pipe
+			m := transport.Message{
+				Header: map[string]string{
+					"Micro-Method": "advert",
+				},
+				Body: body,
+			}
+			if err := client.Send(&m); err != nil {
+				log.Debugf("Network failed to send advert %s: %v", pbAdvert.Id, err)
+				continue
+			}
+		case <-n.closed:
+			return
+		}
+	}
+}
+
+// Connect connects the network
+func (n *network) Connect() error {
+	n.Lock()
+	defer n.Unlock()
+
+	// return if already connected
+	if n.connected {
+		return nil
+	}
+
+	// try to resolve network nodes
+	nodes, err := n.resolveNodes()
+	if err != nil {
+		return err
+	}
+
+	// connect network tunnel
+	if err := n.Tunnel.Connect(); err != nil {
+		return err
+	}
+
+	// initialize the tunnel to resolved nodes
+	n.Tunnel.Init(
+		tunnel.Nodes(nodes...),
+	)
+
+	// dial into ControlChannel to send route adverts
+	client, err := n.Tunnel.Dial(ControlChannel)
+	if err != nil {
+		// TODO: should we stop the tunnel here?
+		return err
+	}
+
+	// create closed channel
+	n.closed = make(chan bool)
+
+	// keep resolving network nodes
+	go n.resolve()
+
+	// TODO: do we assume the router has been started?
+	// start advertising routes
+	advertChan, err := n.options.Router.Advertise()
+	if err != nil {
+		return err
+	}
+
+	// advertise routes
+	go n.advertise(client, advertChan)
+	// process routes
+	go n.process(client)
+
+	// set connected to true
+	n.connected = true
+
+	return nil
+}
+
+func (n *network) close() error {
+	// stop the router
```

Comment:
Stop the server

Reply:
Ditto. Should we stop the router here as well?

Reply:
Yea